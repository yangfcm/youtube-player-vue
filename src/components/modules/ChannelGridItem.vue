<template>
  <div class="ui card centered" style="margin: 0;">
    <router-link :to="`/channel/${channel.snippet.resourceId.channelId}`" class="image">
      <img :src="channel.snippet.thumbnails.medium.url" :alt="channel.snippet.title" />
    </router-link>
    <div class="content" style="overflow: hidden; min-height: 72px">
      <router-link
        class="header ui aligned center"
        :to="`/channel/${channel.snippet.resourceId.channelId}`"
      >{{ channel.snippet.title }}</router-link>
    </div>
  </div>
</template>

<script>
export default {
  props: ["channel"],
};
</script>

<style scoped>
</style>