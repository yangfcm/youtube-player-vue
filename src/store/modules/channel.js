import axios from "../../settings";
import {
  FETCH_MY_CHANNELS,
  FETCH_CHANNEL_SUBSCRIPTION,
  FETCH_CHANNEL_INTRO,
  SUBSCRIBE_CHANNEL,
  UNSUBSCRIBE_CHANNEL,
  CATCH_CHANNEL_ERROR,
} from "../types";
import parseError from "../helpers/parseError";

const state = {
  myChannels: null, // channels subscribed by authed user
  channelIntro: null,
  subscriptionId: null, // Give a channel id, record its subscription id(if subscribed), or empty string(not subscribed)
  channelError: null,
};

const getters = {
  channelErrorMessage: (state) =>
    parseError(state.channelError, "Failed to fetch channel"),
};

const mutations = {
  FETCH_MY_CHANNELS: (state, payload) => {
    state.myChannels = payload;
    state.channelError = null;
  },
  FETCH_CHANNEL_SUBSCRIPTION: (state, payload) => {
    state.subscriptionId = payload;
  },
  FETCH_CHANNEL_INTRO: (state, payload) => {
    state.channelIntro = payload;
    state.channelError = null;
  },
  SUBSCRIBE_CHANNEL: (state, payload) => {
    state.subscriptionId = payload; // If a channel is subscribed successfully, save its subscription id
  },
  UNSUBSCRIBE_CHANNEL: (state) => {
    state.subscriptionId = ""; // If a subscribed channel is unsubscribed, set its subscriptionId as empty string
  },
  CATCH_CHANNEL_ERROR: (state, payload) => {
    state.channelError = payload;
  },
};

const actions = {
  /** Get the channels subscribed by authed user */
  fetchMyChannels: async (context, pageToken) => {
    try {
      const accessToken = localStorage.getItem("access_token");
      const response = await axios.get("/subscriptions", {
        headers: {
          Authorization: accessToken,
        },
        params: {
          ...axios.defaults.params,
          part: "snippet",
          maxResults: 50,
          order: "alphabetical",
          mine: true,
          pageToken,
        },
      });
      context.commit(FETCH_MY_CHANNELS, response.data);
    } catch (err) {
      context.commit(CATCH_CHANNEL_ERROR, err);
    }
  },

  /** Get the channel info by id */
  fetchChannelIntro: async (context, channelId) => {
    try {
      const response = await axios.get("/channels", {
        params: {
          ...axios.defaults.params,
          part: "snippet,statistics",
          id: channelId,
        },
      });
      context.commit(FETCH_CHANNEL_INTRO, response.data);
    } catch (err) {
      context.commit(CATCH_CHANNEL_ERROR, err);
    }
  },

  /** Check if the channel with channel id is subscribed by authed user */
  fetchChannelSubscription: async (context, channelId) => {
    try {
      const accessToken = localStorage.getItem("access_token");
      const response = await axios.get("/subscriptions", {
        headers: {
          Authorization: accessToken,
        },
        params: {
          ...axios.defaults.params,
          part: "snippet",
          forChannelId: channelId,
          mine: true,
        },
      });
      const subscribedChannel = response.data.items[0];
      context.commit(
        FETCH_CHANNEL_SUBSCRIPTION,
        subscribedChannel ? subscribedChannel.id : ""
      );
    } catch (err) {
      context.commit(CATCH_CHANNEL_ERROR, err);
    }
  },

  /** Subscribe a channel */
  subscribeChannel: async (context, channelId) => {
    try {
      const accessToken = localStorage.getItem("access_token");
      const response = await axios.post(
        "/subscriptions",
        {
          snippet: {
            resourceId: {
              kind: "youtube#channel",
              channelId,
            },
          },
        },
        {
          headers: {
            Authorization: accessToken,
          },
          params: {
            part: "snippet",
          },
        }
      );
      context.commit(SUBSCRIBE_CHANNEL, response.data.id);
    } catch (err) {
      context.commit(CATCH_CHANNEL_ERROR, err);
    }
  },

  /** Unsubscribe a channel */
  unsubscribeChannel: async (context, subscriptionId) => {
    try {
      const accessToken = localStorage.getItem("access_token");
      // const { subscriptionId } = context.state;
      await axios.delete("/subscriptions", {
        headers: {
          Authorization: accessToken,
        },
        params: {
          id: subscriptionId,
        },
      });
      context.commit(UNSUBSCRIBE_CHANNEL);
    } catch (err) {
      context.commit(CATCH_CHANNEL_ERROR, err);
    }
  },
};

export default {
  state,
  getters,
  mutations,
  actions,
};
