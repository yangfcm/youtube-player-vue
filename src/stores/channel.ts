import { ref } from 'vue'
import { defineStore } from 'pinia'
import { AsyncStatus } from '@/settings/types'
import { fetchChannelProfileAPI, fetchChannelVideosAPI, fetchChannelPlayListsAPI } from './api'
import type { ChannelDetails, PlayListsRespone, VideosResponse } from './types'

type ChannelStore = {
  profile: {
    status: AsyncStatus
    error: string
    data: Record<string, ChannelDetails>
  }
  videos: {
    status: AsyncStatus
    error: string
    data: Record<string, VideosResponse>
  }
  playlists: {
    status: AsyncStatus
    error: string
    data: Record<string, PlayListsRespone>
  }
}

export const useChannelStore = defineStore('channel', () => {
  const channel = ref<ChannelStore>({
    profile: {
      status: AsyncStatus.IDLE,
      error: '',
      data: {},
    },
    videos: {
      status: AsyncStatus.IDLE,
      error: '',
      data: {},
    },
    playlists: {
      status: AsyncStatus.IDLE,
      error: '',
      data: {},
    },
  })

  const fetchChannelProfile = async (channelId: string) => {
    try {
      channel.value.profile.status = AsyncStatus.LOADING
      channel.value.profile.error = ''
      const response = await fetchChannelProfileAPI(channelId)
      channel.value.profile.status = AsyncStatus.SUCCESS
      channel.value.profile.data[channelId] = response.data.items[0]
    } catch (err: any) {
      channel.value.profile.status = AsyncStatus.FAIL
      channel.value.profile.error = err.message
    }
  }
  const fetchChannelVideos = async (channelId: string, pageToken?: string) => {
    try {
      channel.value.videos.status = AsyncStatus.LOADING
      channel.value.videos.error = ''
      const options: Record<string, string> = {}
      if (pageToken) options.pageToken = pageToken
      const response = await fetchChannelVideosAPI(channelId, options)
      const currentItems = channel.value.videos.data[channelId]?.items || []
      channel.value.videos.data[channelId] = {
        ...response.data,
        items: [...currentItems, ...response.data.items],
      }
      channel.value.videos.status = AsyncStatus.SUCCESS
    } catch (err: any) {
      channel.value.videos.status = AsyncStatus.FAIL
      channel.value.videos.error = err.message
    }
  }
  const fetchChannelPlayLists = async (channelId: string, pageToken?: string) => {
    try {
      channel.value.playlists.status = AsyncStatus.LOADING
      channel.value.playlists.error = ''
      const options: Record<string, string> = {}
      if (pageToken) options.pageToken = pageToken
      const response = await fetchChannelPlayListsAPI(channelId, options)
      const currentItems = channel.value.playlists.data[channelId]?.items || []
      channel.value.playlists.data[channelId] = {
        ...response.data,
        items: [...currentItems, ...response.data.items],
      }
      channel.value.playlists.status = AsyncStatus.SUCCESS
    } catch (err: any) {
      channel.value.playlists.status = AsyncStatus.FAIL
      channel.value.playlists.error = err.message
    }
  }

  return {
    channelState: channel,
    fetchChannelProfile,
    fetchChannelVideos,
    fetchChannelPlayLists,
  }
})
