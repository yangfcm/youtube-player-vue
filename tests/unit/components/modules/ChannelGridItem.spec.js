import { shallowMount, RouterLinkStub } from "@vue/test-utils";
import ChannelGridItem from "@/components/modules/ChannelGridItem.vue";
import { myChannelsResponse } from "../../store/fixtures/channel";

describe("Test ChannelGridItem component", () => {
  let wrapper;
  const channel = myChannelsResponse.items[0];
  beforeEach(() => {
    wrapper = shallowMount(ChannelGridItem, {
      propsData: {
        channel,
      },
      stubs: {
        RouterLink: RouterLinkStub,
      },
    });
  });
  afterEach(() => {
    wrapper.destroy();
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render correct content", () => {
    expect(wrapper.find("img").attributes("src")).toBe(
      channel.snippet.thumbnails.medium.url
    );
    expect(wrapper.find("div.content").text()).toContain(channel.snippet.title);
    // console.log(wrapper.find("a").props());
    expect(wrapper.find("a").props().to).toBe(
      `/channel/${channel.snippet.resourceId.channelId}`
    );
  });
});
