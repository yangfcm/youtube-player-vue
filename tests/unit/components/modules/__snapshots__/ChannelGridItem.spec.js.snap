// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ChannelGridItem component should match snapshot 1`] = `
<div class="ui card centered" style="margin: 0px;"><a class="image"><img src="https://yt3.ggpht.com/-poJuOARXlRM/AAAAAAAAAAI/AAAAAAAAAAA/x83kAYqNPmA/s240-c-k-no-mo-rj-c0xffffff/photo.jpg" alt="Aussie English"></a>
  <div class="content" style="overflow: hidden; min-height: 72px;"><a class="header ui aligned center">Aussie English</a></div>
</div>
`;
